### PR TITLE
Fix abi encoding/decoding

### PR DESCRIPTION
**Pull request by AtkinsChang**
- Fix dynamic array decoding while #142 fixes only encoding part
- Fix fixed-length array encoding/decoding

### CODE REVIEW

**Comment by AtkinsChang** [comment]
[clabot:check]

**Comment by tomusdrw** [comment]
I don't know enough about the ABI spec to be able to review that completely (unless you point me to the details). The code looks nice & clean though.

File: ethabi/src/decoder.rs [hunk 1]
```diff
@@ -189,13 +189,29 @@ fn decode_param(param: &ParamType, slices: &[[u8; 32]], offset: usize) -> Result
...
-			let mut tokens = vec![];
-			let mut new_offset = offset;
-			for _ in 0..len {
-				let res = try!(decode_param(t, &slices, new_offset));
-				new_offset = res.new_offset;
-				tokens.push(res.token);
-			}
+			let mut tokens = Vec::with_capacity(len);
+			let new_offset = if param.is_dynamic() {
+				let offset_slice = try!(peek(slices, offset));
```
Comment (dvdplm) [inline ethabi/src/decoder.rs]: minor nitpick: new code should prefer `?` rather than `try!`

File: ethabi/src/token/token.rs [hunk 2]
```diff
@@ -73,6 +73,15 @@ impl fmt::Display for Token {
 }
 
 impl Token {
+	/// returns wheater this token is dynamic
```
Comment (dvdplm) [inline ethabi/src/token/token.rs]:
```suggestion
	/// Returns whether this token is dynamic. `FixedArray` is considered dynamic if any of its members is dynamic.
```

**Pull request merged**